### ISSUE TITLE
Feature request: sentence-level weights

### ISSUE THREAD

**Issue by kpu** [marian-nmt/marian-dev#171]
@phikoehn is campaigning for a WMT task on data cleaning using the raw data from http://paracrawl.eu/download.html .  Rather than just simple thresholding, I suggest we let people assign weights to sentences.  This could mean weighted objective or floating-point sampling probability.  In the latter case, the definition of epoch becomes somewhat loose.  

If the task comes with a canned Marian system, then it's good marketing because people will use Marian.

**Comment by emjotde** [comment]
Second option, floating-point sampling probability, seems a good way to do that as it does not affect models or code outside the data iterators. I would restrict this functionality to the new sqlite-based corpus container only.

**Comment by phikoehn** [comment]
The paper

Chen, Boxing and Cherry, Colin and Foster, George and Larkin, Samuel (2017): Cost Weighting for Neural Machine Translation Domain Adaptation, Proceedings of the First Workshop on Neural Machine Translation

http://www.aclweb.org/anthology/W17-3205.pdf

proposes to scale the learning rate by a sentence-specific factor. This has the advantage that no low-scoring sentence pairs are skipped.

**Comment by emjotde** [comment]
I guess we can do both independently, the second version can also be used for domain adaptation. I am also tending to abandon the typical concept of epoch as a complete data sweep towards an epoch as a specific number of randomly sampled batches. That would make training more comparable across different data sets.

**Comment by snukky** [comment]
Data weighting (`--data-weighting file.txt`) has been added to the master branch of marian-dev. It may work at sentence (default) and word level (`--data-weighting-type word`).

**Comment by phikoehn** [comment]
Is this using weighted sampling, or learning rate scaling?

**Comment by emjotde** [comment]
All word-level cross-entropy scores get multiplied with a weight before summing them up. Depending on the mode this is the same weight for the whole sentence or different weights per word. The former should be equivalent to learning rate scaling.

**Comment by emjotde** [comment]
@snukky good job, btw.

**Comment by phikoehn** [comment]
Great! Now we only have to find good weights. The one with the most GPUs wins. Oh, well.

**Comment by emjotde** [comment]
@phikoehn You brought that one on yourself :P

**Comment by fiqas** [comment]
I've finished implementing word-level dynamic weighting based on frequencies during the training. I plan to test it now and to refactor the code so that it's easier to add any type of weighting in the future.

**Comment by emjotde** [comment]
Seems to work.

**Comment by GokulNC** [comment]
@emjotde @snukky 
Is there anyway we can use the sentence-weighting as a sampling probability instead of weighting factor that gets multiplied with the cost?

Use-case: I have a corpus with 10M pure translated data and 200M back-translated data.
I do not want each epoch to train on all sentences from back-translated data.

So is it somehow possible to set probability of sampling a sentence for the training corpus?

Thanks!
